File: examples/sdk_example.js
/**
 * Example: Using the Promptware Node.js SDK
 *
 * Demonstrates all SDK features:
 * - Dynamic verb discovery
 * - Automatic retries
 * - Circuit breaker
 * - Health checks
 * - Error handling
 */

import {
  Agent,
  callVerb,
  VerbNotFoundError,
  InvalidParamsError,
  ConnectionError,
  CircuitBreakerError,
} from '../promptware-js/sdk.js';

async function basicUsage() {
  console.log('=== Basic Usage ===\n');

  // Create agent client
  const agent = new Agent('http://localhost:3000');

  // Health check
  const health = await agent.health();
  console.log('Health:', health);

  // List available verbs
  const verbs = await agent.listVerbs();
  console.log('\nAvailable verbs:', verbs);

  // Call verb with dot notation (assumes @v1)
  try {
    const result = await agent.user.create({
      email: 'alice@example.com',
      name: 'Alice Johnson'
    });
    console.log('\nUser created:', result);
  } catch (error) {
    if (error instanceof VerbNotFoundError) {
      console.log('\nNote: user.create@v1 verb not found - using mock agent');
    }
  }
}

async function errorHandling() {
  console.log('\n=== Error Handling ===\n');

  const agent = new Agent('http://localhost:3000');

  // Verb not found
  try {
    await agent.nonexistent.verb();
  } catch (error) {
    if (error instanceof VerbNotFoundError) {
      console.log('Verb not found:', error.message);
    }
  }

  // Invalid parameters
  try {
    await agent.user.create();  // Missing required params
  } catch (error) {
    if (error instanceof InvalidParamsError) {
      console.log('Invalid parameters:', error.message);
    }
  }

  // Connection error
  try {
    const badAgent = new Agent('http://localhost:9999', { timeout: 5000 });
    await badAgent.health();
  } catch (error) {
    if (error instanceof ConnectionError) {
      console.log('Connection failed:', error.message);
    }
  }
}

async function withRetries() {
  console.log('\n=== Automatic Retries ===\n');

  const agent = new Agent('http://localhost:3000', {
    maxRetries: 5,
    retryDelay: 2000,
    timeout: 60000
  });

  // Will automatically retry on network errors
  try {
    const result = await agent.user.get({ userId: '123' });
    console.log('User retrieved (with retries):', result);
  } catch (error) {
    if (error instanceof VerbNotFoundError) {
      console.log('Note: user.get@v1 verb not found - using mock agent');
    }
  }
}

async function circuitBreakerExample() {
  console.log('\n=== Circuit Breaker ===\n');

  const agent = new Agent('http://localhost:3000', {
    circuitBreakerThreshold: 3,  // Open after 3 failures
    circuitBreakerTimeout: 10000  // Try again after 10 seconds
  });

  // Listen to circuit breaker events
  agent.on('circuit-breaker-state', (state) => {
    console.log(`[Circuit Breaker] State: ${state}`);
  });

  // Simulate failures
  const badAgent = new Agent('http://localhost:9999', {
    circuitBreakerThreshold: 2,
    timeout: 2000
  });

  for (let i = 0; i < 5; i++) {
    try {
      await badAgent.health();
    } catch (error) {
      if (error instanceof CircuitBreakerError) {
        console.log(`Attempt ${i+1}: Circuit breaker is OPEN - service unavailable`);
        break;
      } else if (error instanceof ConnectionError) {
        console.log(`Attempt ${i+1}: Connection failed`);
      }
    }
  }
}

async function verbDiscovery() {
  console.log('\n=== Verb Discovery ===\n');

  const agent = new Agent('http://localhost:3000');

  // Discover all verbs
  const verbs = await agent.discover();
  console.log(`Discovered ${verbs.length} verbs:`);

  for (const verb of verbs) {
    console.log(`\nVerb: ${verb.name}`);
    console.log(`  Description: ${verb.description || 'N/A'}`);

    // Show parameters
    const inputSchema = verb.inputSchema || {};
    const properties = inputSchema.properties || {};
    const required = inputSchema.required || [];

    if (Object.keys(properties).length > 0) {
      console.log('  Parameters:');
      for (const [paramName, paramSchema] of Object.entries(properties)) {
        const requiredStr = required.includes(paramName) ? ' (required)' : '';
        const paramType = paramSchema.type || 'unknown';
        console.log(`    - ${paramName}: ${paramType}${requiredStr}`);
      }
    }
  }
}

async function productionConfig() {
  console.log('\n=== Production Configuration ===\n');

  const agent = new Agent('http://localhost:3000', {
    timeout: 60000,
    maxRetries: 5,
    retryDelay: 2000,
    retryBackoff: 2.0,
    circuitBreakerThreshold: 10,
    circuitBreakerTimeout: 120000,
    enableLogging: true
  });

  // Monitor circuit breaker
  agent.on('circuit-breaker-state', (state) => {
    console.log(`[Production] Circuit breaker state: ${state}`);
  });

  // Use in production with proper error handling
  try {
    // Health check before processing
    const health = await agent.health();
    if (health.status !== 'alive') {
      throw new Error('Service not healthy');
    }

    // Process requests
    const result = await agent.user.create({
      email: 'production@example.com',
      name: 'Production User'
    });
    console.log('Production request successful:', result);

  } catch (error) {
    if (error instanceof VerbNotFoundError) {
      console.log('Note: user.create@v1 verb not found - using mock agent');
    } else {
      console.log('Production error:', error.message);
    }
  } finally {
    agent.close();
  }
}

async function convenienceFunction() {
  console.log('\n=== Convenience Function ===\n');

  // One-off verb call
  try {
    const result = await callVerb(
      'http://localhost:3000',
      'user.get@v1',
      { userId: '123' }
    );
    console.log('One-off call result:', result);
  } catch (error) {
    if (error instanceof VerbNotFoundError) {
      console.log('Note: user.get@v1 verb not found - using mock agent');
    }
  }
}

// Main execution
(async () => {
  console.log('Promptware SDK Examples\n');
  console.log('Make sure you have an agent running on http://localhost:3000');
  console.log('Or use: promptware generate test.pw --lang python && cd generated/test && python test_server.py\n');

  try {
    await basicUsage();
    await errorHandling();
    await withRetries();
    await circuitBreakerExample();
    await verbDiscovery();
    await productionConfig();
    await convenienceFunction();

    console.log('\n=== All Examples Complete ===');
  } catch (error) {
    console.error('\nUnexpected error:', error);
  }
})();

File: examples/sdk_example.py
"""
Example: Using the Promptware Python SDK

Demonstrates all SDK features:
- Dynamic verb discovery
- Automatic retries
- Circuit breaker
- Health checks
- Error handling
"""

import logging

from assertlang.sdk import (
    Agent,
    CircuitBreakerError,
    ConnectionError,
    InvalidParamsError,
    VerbNotFoundError,
)

# Enable logging to see what's happening
logging.basicConfig(level=logging.INFO)


def basic_usage():
    """Basic SDK usage example."""
    print("=== Basic Usage ===\n")

    # Create agent client
    agent = Agent("http://localhost:3000")

    # Health check
    health = agent.health()
    print(f"Health: {health}")

    # List available verbs
    verbs = agent.list_verbs()
    print(f"\nAvailable verbs: {verbs}")

    # Call verb with dot notation (assumes @v1)
    try:
        result = agent.user.create(
            email="alice@example.com",
            name="Alice Johnson"
        )
        print(f"\nUser created: {result}")
    except VerbNotFoundError:
        print("\nNote: user.create@v1 verb not found - using mock agent")


def with_context_manager():
    """Using context manager for automatic cleanup."""
    print("\n=== Context Manager ===\n")

    with Agent("http://localhost:3000") as agent:
        # Connection automatically closed when exiting
        ready = agent.ready()
        print(f"Readiness: {ready}")


def error_handling():
    """Comprehensive error handling."""
    print("\n=== Error Handling ===\n")

    agent = Agent("http://localhost:3000")

    try:
        # This will fail if verb doesn't exist
        agent.nonexistent.verb()
    except VerbNotFoundError as e:
        print(f"Verb not found: {e}")

    try:
        # This will fail if parameters are invalid
        agent.user.create()  # Missing required params
    except InvalidParamsError as e:
        print(f"Invalid parameters: {e}")

    try:
        # This will fail if service is down
        bad_agent = Agent("http://localhost:9999", timeout=5)
        bad_agent.health()
    except ConnectionError as e:
        print(f"Connection failed: {e}")


def with_retries():
    """Automatic retry configuration."""
    print("\n=== Automatic Retries ===\n")

    agent = Agent(
        "http://localhost:3000",
        max_retries=5,
        retry_delay=2.0,
        timeout=60
    )

    # Will automatically retry on network errors
    try:
        result = agent.user.get(user_id="123")
        print(f"User retrieved (with retries): {result}")
    except VerbNotFoundError:
        print("Note: user.get@v1 verb not found - using mock agent")


def circuit_breaker_example():
    """Circuit breaker pattern."""
    print("\n=== Circuit Breaker ===\n")

    Agent(
        "http://localhost:3000",
        circuit_breaker_threshold=3,  # Open after 3 failures
        circuit_breaker_timeout=10    # Try again after 10 seconds
    )

    # Simulate failures
    bad_agent = Agent(
        "http://localhost:9999",  # Bad URL
        circuit_breaker_threshold=2,
        timeout=2
    )

    for i in range(5):
        try:
            bad_agent.health()
        except CircuitBreakerError:
            print(f"Attempt {i+1}: Circuit breaker is OPEN - service unavailable")
            break
        except ConnectionError:
            print(f"Attempt {i+1}: Connection failed")


def verb_discovery():
    """Discover and introspect verbs."""
    print("\n=== Verb Discovery ===\n")

    agent = Agent("http://localhost:3000")

    # Discover all verbs
    verbs = agent.discover()
    print(f"Discovered {len(verbs)} verbs:")

    for verb in verbs:
        print(f"\nVerb: {verb['name']}")
        print(f"  Description: {verb.get('description', 'N/A')}")

        # Show parameters
        input_schema = verb.get('inputSchema', {})
        properties = input_schema.get('properties', {})
        required = input_schema.get('required', [])

        if properties:
            print("  Parameters:")
            for param_name, param_schema in properties.items():
                required_str = " (required)" if param_name in required else ""
                param_type = param_schema.get('type', 'unknown')
                print(f"    - {param_name}: {param_type}{required_str}")


def production_config():
    """Production-ready configuration."""
    print("\n=== Production Configuration ===\n")

    agent = Agent(
        base_url="http://localhost:3000",
        timeout=60,
        max_retries=5,
        retry_delay=2.0,
        circuit_breaker_threshold=10,
        circuit_breaker_timeout=120,
        enable_logging=True
    )

    # Use in production with proper error handling
    try:
        with agent:
            # Health check before processing
            health = agent.health()
            if health.get('status') != 'alive':
                raise RuntimeError("Service not healthy")

            # Process requests
            result = agent.user.create(
                email="production@example.com",
                name="Production User"
            )
            print(f"Production request successful: {result}")

    except VerbNotFoundError:
        print("Note: user.create@v1 verb not found - using mock agent")
    except Exception as e:
        print(f"Production error: {e}")


def convenience_function():
    """Using the convenience function for one-off calls."""
    print("\n=== Convenience Function ===\n")

    from assertlang.sdk import call_verb

    # One-off verb call
    try:
        result = call_verb(
            base_url="http://localhost:3000",
            verb_name="user.get@v1",
            params={"user_id": "123"}
        )
        print(f"One-off call result: {result}")
    except VerbNotFoundError:
        print("Note: user.get@v1 verb not found - using mock agent")


if __name__ == "__main__":
    print("Promptware SDK Examples\n")
    print("Make sure you have an agent running on http://localhost:3000")
    print("Or use: promptware generate test.al --lang python && cd generated/test && python test_server.py\n")

    try:
        basic_usage()
        with_context_manager()
        error_handling()
        with_retries()
        circuit_breaker_example()
        verb_discovery()
        production_config()
        convenience_function()

        print("\n=== All Examples Complete ===")

    except KeyboardInterrupt:
        print("\n\nExamples interrupted by user")
